### DISCUSSION TITLE
Binja features

### DISCUSSION THREAD

**Discussion by b00tkitism** [NationalSecurityAgency/ghidra#5562] [Q&A] [answered]
Do you have any plans to add some powerful features like SSA and your own IL(Intermediate Language) like Binary Ninja ?

**Comment by d-millar** [comment, marked as answer]
@alom4hdi Ghidra has had its own SSA IL (which we call PCode) since its public release and for many years prior.  It is documented in GhidraDocs/languages and the class materials provided with the distribution.

**Reply by LukeSerne** [reply]
Unfortunately, the low p-code that is generated from the assembly instructions (and can be shown in the listing in the "PCode" field) is not in SSA form (for example, register names are used as-is and can thus be assigned to multiple times). Transforming this p-code into SSA form is done by the decompiler in the `heritage` stage. It is possible to read an SSA form of p-code from the java side (namely the high pcode), but that also includes the other optimisations the decompiler performs.

**Comment by b00tkitism** [comment]
Nice, Thanks.
I'm a beginner software reverse engineer, and I've used IDA and Binja before.

Today I tried to reverse engineer a Windows kernel level driver for NDIS, IDA Pro and Binja detected It's functions and types very well but Ghidra doesn't detect them (e.g. NDIS_HANDLE).
I used default settings for analyzing and decompiling. 
what should I change to get it working ?

**Reply by d-millar** [reply]
@alom4hdi You'll probably need to parse the header files that contain the NDIS types to produce the requisite data-type archives, assuming you don't have debug symbols for them.  I am probably not the guy to walk you through that, but will ping the folks who do it regularly.  There is some documentation i the distro for doing this - just look for CParser.

**Reply by rvrecen** [reply]
You can parse C header files in batch via this menu

![image](https://github.com/NationalSecurityAgency/ghidra/assets/66631929/8ceb9e40-0051-4c16-a128-c3d15797c469)

Assuming parsing C header is what is wanted here. This will load in the data-type sigs and I believe function signatures. This would require the source code needed.

There is also a chance what you want/need already exists in uploaded GDTs
https://github.com/AllsafeCyberSecurity/Ghidra_Data_Type

**Comment by b00tkitism** [comment]
and another question
does Ghidra have something similar to Binja's multi level IL ? (LLIL, MLIL)

**Reply by d-millar** [reply]
Sort of, but not really - we have low pcode (related to the disassembly) and high pcode (related to the decompilation) and there are several variants of the high pcode.  These are similar but not truly equivalent to the Binja IL system.

**Reply by b00tkitism** [reply]
It will be great to have something like Binja

**Comment by d-millar** [comment]
@alom4hdi What specifically are you looking for?  I.e. what is it you can do in Binja that you are unable to do in Ghidra (or even find it easier to do in Binja)?

**Reply by b00tkitism** [reply]
Some of them:
 - LLIL, MLIL(understanding assembly better with reading ILs step by step)
 - Better decompilation I think(correct me If I'm wrong)
 - Rust API

**Reply by rvrecen** [reply]
In the set register value menu, it would be neat to cover similar things as Binja in Ghidra if the auto could utilize it.

![image](https://github.com/NationalSecurityAgency/ghidra/assets/66631929/40b5c19f-5c73-43f7-b03d-d61df0054658)

Unrelated the fact this says unknown is a little surprising
![image](https://github.com/NationalSecurityAgency/ghidra/assets/66631929/b90e00d9-4dc6-4fa6-b3c9-a418d2ec0a1f)

vs
![image](https://github.com/NationalSecurityAgency/ghidra/assets/66631929/b7b8760e-7397-4863-becf-2ec7fb659225)

I am not familiar with varnodes to know whether this is possible, and in Binja land this is in their IL only, not disassembly.

I don't think the constant prop currently handles VSA, and that sounds like a major feature change.

**Comment by Mandofskii** [comment]
Wish you implement an intermediate language for Ghidra like BNIL.
It's so useful for understanding what's going on in a binary.

**Reply by d-millar** [reply]
As noted above, there already IS an intermediate language for Ghidra.   If you wish to see an example, simply enable the field called "PCode" in the Code Browser.

**Reply by Mandofskii** [reply]
I mean a better version of it(multi levels like what Binary Ninja did) for example in Binary Ninja you can use LLIL APIs to find out how many times eax reassigned in a single function.

![BNIL](https://github.com/NationalSecurityAgency/ghidra/assets/59086060/ec0f8bd7-c8f3-42bb-9612-e59561b04416)

**Reply by d-millar** [reply]
Ah, understood!

**Reply by Mandofskii** [reply]
Do you have any idea for implementing this ?

I need some hints to start its development.

**Reply by d-millar** [reply]
Wow, that's a pretty big task to take on.   Best hint I can offer is don't, sadly.

**Reply by astrelsky** [reply]
> Wow, that's a pretty big task to take on. Best hint I can offer is don't, sadly.

This may be the most solid piece of advice I've seen you give (and you give quite a lot).
